### PR TITLE
Simplify existing paypal calls and add support for new preapproval flow

### PR DESCRIPTION
**Pull request by asood123**
Under the new paypal setup, we don't need to worry about anything except expiry date of a preapproval key. So, this PR records start and end date of the keys separately in a table row.

### CODE REVIEW

File: migrations/20170228184811-add-fields-to-PaymentMethod-migration.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+module.exports = {
+  up: function (queryInterface, Sequelize) {
+    return queryInterface.addColumn('PaymentMethods', 'startDate', {
```
Comment (xdamman) [inline migrations/20170228184811-add-fields-to-PaymentMethod-migration.js]: why a startDate? Isn't it always the same as `createdAt`?

File: migrations/20170228184811-add-fields-to-PaymentMethod-migration.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      type: Sequelize.DATE
+    })
+    .then(() => queryInterface.addColumn('PaymentMethods', 'endDate', {
```
Comment (xdamman) [inline migrations/20170228184811-add-fields-to-PaymentMethod-migration.js]: wouldn't it make more sense to call this `expiryDate`?

File: server/controllers/paypal.js [hunk 3]
```diff
@@ -16,193 +15,121 @@ const {
 export const getDetails = function(req, res, next) {
   const preapprovalKey = req.params.preapprovalkey;
 
-  return paypalAdaptive.preapprovalDetails(preapprovalKey)
-    .then(paypalResponse => res.json(paypalResponse))
+  return getPreapprovalDetailsAndUpdatePaymentMethod(preapprovalKey, req.remoteUser.id)
+    .then(response => res.json(response))
```
Comment (xdamman) [inline server/controllers/paypal.js]: `.then(res.json)`

File: server/controllers/paypal.js [hunk 4]
```diff
@@ -16,193 +15,121 @@ const {
...
+  
+  const payload = {
+    currencyCode: 'USD', // TODO: figure out if there is a reliable way to specify correct currency for a HOST.
+    startingDate: startDate.toISOString(),
+    endingDate: endDate.toISOString(),
+    returnUrl,
+    cancelUrl,
+    displayMaxTotalAmount: false,
+    feesPayer: 'SENDER',
+    maxAmountPerPayment: 2500.00, // PayPal claims this can go up to $10k without needing additional permissions from them.
```
Comment (xdamman) [inline server/controllers/paypal.js]: so should we raise this up to $4-5k?

File: server/controllers/paypal.js [hunk 5]
```diff
@@ -16,193 +15,121 @@ const {
...
+    Promise.reject(new errors.NotFound('This preapprovalKey does not exist.')))
+
+  .then(() => getPreapprovalDetailsAndUpdatePaymentMethod(req.params.preapprovalkey, req.remoteUser.id, paymentMethod))
+
+  .then(() => Activity.create({
+    type: 'user.paymentMethod.created',
+    UserId: req.remoteUser.id,
+    data: {
+      user: req.remoteUser.minimal,
+      paymentMethod
```
Comment (xdamman) [inline server/controllers/paypal.js]: not sure it's safe to store the paymentMethod in the activity. We should sanitize it first.

File: server/controllers/paypal.js [hunk 6]
```diff
@@ -16,193 +15,121 @@ const {
...
-            token: req.params.preapprovalkey
-          }
-        })
-        .then(paymentMethod => cb(null, paymentMethod))
-        .catch(cb);
-    }],
+/*
+ * Takes a Payment Response
+ */
+const getPreapprovalDetailsAndUpdatePaymentMethod = function(preapprovalKey, userId, paymentMethod = null) {
```
Comment (xdamman) [inline server/controllers/paypal.js]: why not splitting this in two? `getPreaprovalDetails` and `updatePaymentMethod` ?

File: server/controllers/paypal.js (hunk 4 above)
Comment (asood123) [inline server/controllers/paypal.js]: Let's start lower and make sure it works first. We can easily disable old preapprovals and have them go through the flow again on website now.

File: server/controllers/paypal.js (hunk 5 above)
Comment (asood123) [inline server/controllers/paypal.js]: Activity isn't used to share anything publicly - it's full of payments data, everything from stripe tokens to paypal keys. Though this should be .info, i'll change it.

File: server/controllers/paypal.js (hunk 6 above)
Comment (asood123) [inline server/controllers/paypal.js]: The two always happen at the same time. Whenever we fetch details, we need to update the paymentMethod. If needed, we can split in the future.

File: server/controllers/paypal.js (hunk 3 above)
Comment (asood123) [inline server/controllers/paypal.js]: changed

**Comment by xdamman** [comment]
I'd recommend not having those sensitive data in more than one place for
security reasons.

File: migrations/20170228184811-add-fields-to-PaymentMethod-migration.js (hunk 1 above)
Comment (asood123) [inline migrations/20170228184811-add-fields-to-PaymentMethod-migration.js]: removed

**Comment by asood123** [comment]
Merging, tests passed on circleci.

File: server/controllers/paypal.js (hunk 3 above)
Comment (asood123) [inline server/controllers/paypal.js]: fyi, this doesn't work

**Pull request merged**